### PR TITLE
Bug 1857215 - Part 1:Integrate review checker recommendations

### PR DESCRIPTION
**Pull request by rahulsainani**
### What
First step to integrate the product recommendations API

### How
– Integrate the exposed AC API
– Use the first item of the list if present or return null
– Extract `EnumMapper` since it's used in multiple places and will be used for product recommendations mapping as well.

### What's next
Add actions, mapping and middleware  logic.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1857215

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/EnumMapper.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.shopping.middleware
+
+/**
+ * Converts a string to an enum value, ignoring case. If the string does not match any of the
+ * enum values, the default value is returned.
+ */
+inline fun <reified T : Enum<T>> String.asEnumOrDefault(defaultValue: T? = null): T? =
+    enumValues<T>().firstOrNull { it.name.equals(this, ignoreCase = true) } ?: defaultValue
```
Comment (calumozilla) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/EnumMapper.kt:12]: are these enumMapper changes related to recommendations? should they be in this PR

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt [hunk 2]
```diff
@@ -43,6 +44,13 @@ interface ReviewQualityCheckService {
      * Returns the selected tab url.
      */
     fun selectedTabUrl(): String?
+
+    /**
+     * Fetches product recommendations related to the product user is browsing in the current tab.
+     *
+     * @return [ProductRecommendation] if request succeeds, null otherwise.
+     */
+    suspend fun productRecommendations(): ProductRecommendation?
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:53]: the plurality/singularity verbiage here is mismatched. The API returns a list of recommendations, but based on the default implementation, we only ever want the first one that's returned?

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt [hunk 3]
```diff
@@ -110,8 +118,23 @@ class DefaultReviewQualityCheckService(
...
-    private inline fun <reified T : Enum<T>> String.asEnumOrDefault(defaultValue: T? = null): T? =
-        enumValues<T>().firstOrNull { it.name.equals(this, ignoreCase = true) } ?: defaultValue
+    override suspend fun productRecommendations(): ProductRecommendation? =
+        withContext(Dispatchers.Main) {
+            suspendCoroutine { continuation ->
+                browserStore.state.selectedTab?.let { tab ->
+                    tab.engineState.engineSession?.requestProductRecommendations(
+                        url = tab.content.url,
+                        onResult = {
+                            continuation.resume(it.firstOrNull())
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:128]: let's include a comment why we only care about the first recommendation and ignoring whatever else the API returns

**Review by MozillaNoah — APPROVED** [review]

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt (hunk 2 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:53]: Indeedio, it was returning a list earlier and then middleware was responsible for the logic of getting the first one but makes sense to do it here at the edge since we only care about the first recommendation.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/EnumMapper.kt (hunk 1 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/EnumMapper.kt:12]: Yes, since this will be used in the recommendations mapping logic (next PR) along with the 3 places it's used currently, the idea was to split this bug into 2 PRs(commits) in effort to make it easier for reviews, so the second part won't have these changes of extracting the enum mapper and tests.
